src/test/java/org/mifos/integrationtest/cucumber/stepdef/BillPayStepDef.java

```diff
@@ -309,4 +323,195 @@ public void iCanExtractTheCallbackBodyAndAssertTheRtpStatus() {
+                    }
+                    assertThat(code).isNotEmpty();
+                }
+            }
+
+        }
+        assertThat(flag).isTrue();
+    }
+
+    @Then("I should be able to extract response body from callback for bill notification with missing values")
+    public void iShouldBeAbleToExtractResponseBodyFromCallbackForBillNotificationWithMissingValues() throws JSONException {
+        JSONObject jsonObject = new JSONObject(scenarioScopeState.response);
+        scenarioScopeState.transactionId = jsonObject.getString("transactionId");
+        assertThat(scenarioScopeState.transactionId.equals("Invalid Request: Mandatory Fields Missing")).isTrue();
+    }
+
+    @Then("I should be able to extract response body from callback for bill already paid")
+    public void iShouldBeAbleToExtractResponseBodyFromCallbackForBillAlreadyPaid() {
+        boolean flag = false;
+        List<ServeEvent> allServeEvents = getAllServeEvents();
+        for (int i = allServeEvents.size() - 1; i >= 0; i--) {
+            ServeEvent request = allServeEvents.get(i);
+            if (!(request.getRequest().getBodyAsString()).isEmpty()) {
+                JsonNode rootNode = null;
+                flag = true;
+                try {
+                    rootNode = objectMapper.readTree(request.getRequest().getBody());
+                    logger.info("Rootnode value:" + rootNode);
+                } catch (IOException e) {
+                    throw new RuntimeException(e);
+                }
+                if (rootNode != null && rootNode.has("code") && rootNode.get("code").asText().equals("01")) {
+                    String reason = null;
+                    if (rootNode.has("reason")) {
+                        reason = rootNode.get("reason").asText();
+                    }
+                    assertThat(reason).isNotEmpty();
+                    assertThat(reason).contains("Bill Payment Failed: Bill Already Paid");
+                    String code = null;
+                    if (rootNode.has("code")) {
+                        code = rootNode.get("code").asText();
+                    }
+                    assertThat(code).isNotEmpty();
+                }
+            }
+
+        }
+        assertThat(flag).isTrue();
+    }
+
+    @Then("I should not get a response from callback for bill")
+    public void iShouldNotBeAbleToExtractResponseBodyFromCallbackForBill() {
+        boolean flag = false;
+        List<ServeEvent> allServeEvents = getAllServeEvents();
+        assertThat(allServeEvents.size()).isEqualTo(0);
+    }
+
+    @Then("I should be able to extract response body from callback for bill paid after timeout")
+    public void iShouldBeAbleToExtractResponseBodyFromCallbackForBillPaidAfterTimeout() {
+        boolean flag = false;
+        List<ServeEvent> allServeEvents = getAllServeEvents();
+        for (int i = allServeEvents.size() - 1; i >= 0; i--) {
+            ServeEvent request = allServeEvents.get(i);
```

Comment:
Looks like this code is repeated all over, can we use a single method for finding in events and reuse them?

Reply:
https://fynarfin.atlassian.net/browse/PHEE-559